config/config.go

```diff
@@ -1,16 +1,70 @@
 package config
 
-import "time"
+import (
+	ldbopt "github.com/syndtr/goleveldb/leveldb/opt"
+)
 
-// ReplicaConfig specifies the information needed for a Replica and Version manager to
-// maintain global object versioning and provenance.
-// TODO: this configuration is pulled from a service context not a library context.
+// DefaultConfig is used if the user does not specify a configuration
+var DefaultConfig = Config{
+	Versions: ReplicaConfig{
+		PID:    1,
+		Region: "local",
+		Name:   "localhost",
+	},
+}
+
+// New creates a configuration with the required options and can also be used to specify
+// optional configuration e.g. for engine-specific operations.
+func New(options ...Option) Config {
+	// Create the default configuration in editable mode
+	conf := &Config{
+		Versions: ReplicaConfig{
+			PID:    DefaultConfig.Versions.PID,
+			Region: DefaultConfig.Versions.Region,
+			Name:   DefaultConfig.Versions.Name,
+		},
+	}
+
+	// Apply all options to the configuration
+	for _, opt := range options {
+		opt(conf)
+	}
+
+	// Return the value of the configuration
+	return *conf
+}
+
+// Config specifies the options necessary to open a Honu database.
+type Config struct {
+	Versions   ReplicaConfig
+	LDBOptions *ldbopt.Options
+}
+
+// ReplicaConfig specifies the information needed for the Version manager to maintain
+// global object versioning and provenance. This is closely tied to a Replica's
+// configuration, e.g. the PID is the process ID of a running replica, the region is
+// where the replica is running, and the name is usually the hostname of the replica.
 type ReplicaConfig struct {
-	Enabled        bool          `split_words:"true" default:"true"`
-	BindAddr       string        `split_words:"true" default:":4435"`
-	PID            uint64        `split_words:"true" required:"false"`
-	Region         string        `split_words:"true" required:"false"`
-	Name           string        `split_words:"true" required:"false"`
-	GossipInterval time.Duration `split_words:"true" default:"1m"`
-	GossipSigma    time.Duration `split_words:"true" default:"5s"`
+	PID    uint64 `split_words:"true" required:"false"`
+	Region string `split_words:"true" required:"false"`
+	Name   string `split_words:"true" required:"false"`
+}
+
+// Option modifies a configuration to add optional configuration items.
+type Option func(*Config)
```

Comment:
Might some options be mutually exclusive or at least ambiguous in combination?

Reply:
Yes, if you supplied both LevelDB and Pebble configurations, right now it's only possible to use one of them, and both would be ignored if you opened a Badger database. 

Pardon me for making the assumption, but it seems like my answer would prompt you to suggest that we return an error in the case of invalid or ambiguous options? I originally omitted the error return here because I thought people would be doing something to the effect of:

```go
conf := config.New(config.WithLevelDB(opts), config.WithReplica(…))
if db, err = honu.New("leveldb:///db", conf); err != nil {
    …
}
```

I didn't want the error to happen in `config.New` but rather in `honu.New` … we have moved passed that now, and the configuration is as follows:

```
if db, err = honu.New("leveldb:///db", config.WithReplica(…), config.WithLevelDB(opts)); err != nil {
    …
}
```

So if my assumption was correct, then I completely agree with you that we should at least have the ability to return an error and validate the configuration in `honu.New`. I'll make that change!